p2p/src/main/java/bisq/network/p2p/storage/persistence/SplitStoreService.java

```diff
@@ -0,0 +1,190 @@
+import bisq.common.storage.FileUtil;
+import bisq.common.storage.ResourceNotFoundException;
+import bisq.common.storage.Storage;
+
+import java.nio.file.Paths;
+
+import java.io.File;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Has business logic to operate data stores which are spread across multiple files.<br><br><p>
+ *
+ * <h3>Use Case</h3>
+ * Startup requires to send all object keys to the network in order to get the new data.
+ * However, data stores got quite big over time and will grow even faster in the future.
+ * With multi-file data stores, we can query for new objects since the last snapshot (ie.
+ * the last release) and subsequently send "I am Bisq vx.y.z and I have these new objects"
+ * to the network and the network can respond accordingly.</p><br><br><p>
+ *
+ * <h3>Features</h3>
+ * In order to only get a specific part of all the objects available in the complete data
+ * store, the original single-file data store had to be split up and this class and
+ * {@link SplitStore} are there to handle the business logic needed for
+ * <ul>
+ * <li>migrating to the new and shiny multi-file data store</li>
+ * <li>shoveling around data in case Bisq gets updated to a new version</li>
+ * <li>takes care of setting up a fresh Bisq install</li>
+ * <li>makes sure that historical data cannot be altered easily</li>
+ * <li>adding data to the store will only amend the live store</li>
+ * <li>takes care of keeping the legacy API functional</li>
+ * <li>adds the feature of filtering object queries by Bisq release</li>
+ * </ul>
+ * </p>
+ *
+ * <h3>Further reading</h3><p><ul>
+ *     <li><a href="https://github.com/bisq-network/projects/issues/25">project description</a></li>
+ * </ul></p>
+ */
+@Slf4j
+public abstract class SplitStoreService<T extends SplitStore> extends MapStoreService<T, PersistableNetworkPayload> {
+    protected HashMap<String, SplitStore> history;
+
+    public SplitStoreService(File storageDir, Storage<T> storage) {
+        super(storageDir, storage);
+    }
+
+    @Override
+    protected void put(P2PDataStorage.ByteArray hash, PersistableNetworkPayload payload) {
+        // make sure we do not add data that we already have (in a bin of historical data)
+        if (getMap().containsKey(hash))
+            return;
+
+        store.getMap().put(hash, payload);
+        persist();
+    }
+
+    @Override
+    protected PersistableNetworkPayload putIfAbsent(P2PDataStorage.ByteArray hash,
+                                                    PersistableNetworkPayload payload) {
+        // make sure we do not add data that we already have (in a bin of historical data)
+        if (getMap().containsKey(hash))
+            return null;
+
+        PersistableNetworkPayload previous = store.getMap().putIfAbsent(hash, payload);
+        persist();
```

Comment:
agreed. However, it is not trivial to find whether there has been something added to the map or not since the nested `put(...)` only returns the old value.

I believe that is why MapStoreService does this exactly that way (and that might be one source of our IO issues.

I decided to stick with the "proven" way of doing things until we can address the whole disk IO issue in a more general way https://github.com/bisq-network/projects/issues/29